api/v1/controllers/collectors.js

```diff
@@ -159,44 +214,91 @@ function deregisterCollector(req, res, next) {
  * @param {Function} next - The next middleware function in the stack
  */
 function heartbeat(req, res, next) {
-  // TODO reject if caller's token is not a collector token
+  const authToken = req.headers.authorization;
+  const timestamp = req.body.timestamp;
+  let keyName;
+  let tokenName;
   const retval = {
     collectorConfig: config.collector,
     generatorsAdded: [],
     generatorsDeleted: [],
     generatorsUpdated: [],
   };
 
-  /*
-   * TODO update the lastHeartbeat column for this collector
-   */
-
-  /*
-   * TODO Populate collectorConfig
-   * - look up any changes made to this collector since the last heartbeat
-   */
-
-  /*
-   * TODO populate generatorsAdded
-   * - look up any new generators assigned to this collector since the last
-   *   heartbeat
-   * - reEncrypt context values using reEncryptSGContextValues function
-   */
-
-  /*
-   * TODO populate generatorsDeleted
-   * - look up any generators UNassigned from this collector since the last
-   *   heartbeat
-   */
-
-  /*
-   * TODO populate generatorsUpdated
-   * - for generators which were already assigned to this collector, look up
-   *   any changes made to the generator since the last heartbeat
-   * - reEncrypt context values using reEncryptSGContextValues function
-   */
-
-  res.status(httpStatus.OK).json(retval);
+  //verify token
+  jwtUtil.verifyCollectorToken(req)
+  .then(() => jwtUtil.getTokenDetailsFromRequest(req))
+  .then((details) => tokenName = details.username)
+
+  //find collector object
+  .then(() => u.findByKey(helper, req.swagger.params))
+  .then((o) => {
+    keyName = o.name;
+
+    //validate collector
+    if (keyName !== tokenName) {
+      throw new apiErrors.ForbiddenError({
+        explanation: 'Token does not match the specified collector',
+      });
+    } else if (o.status !== 'Running' && o.status !== 'Paused') {
+      throw new apiErrors.ForbiddenError({
+        explanation: `Collector must be running or paused. Status: ${o.status}`,
+      });
+    }
+
+    //setup retval
+    if (heartbeatUtils.collectorMap[o.name]) {
```

Comment:
Wondering, if the copy of "collectorMap" that this file has access to, will get the updates made by the generator patch/post requests. I am not very sure but we might have to do something like we did for the "config" object in the refocus-collector repo.

Reply:
It does. It works fine in the tests.